zio-http/src/main/scala/zhttp/http/middleware/HttpMiddlewares.scala

```diff
@@ -0,0 +1,196 @@
+   */
+  def patchZIO[R, E](f: (Status, Headers) => ZIO[R, Option[E], Patch]): HttpMiddleware[R, E] =
+    makeResponseZIO((_, _, _) => ZIO.unit)((status, headers, _) => f(status, headers))
+
+  /**
+   * Removes the header by name
+   */
+  def removeHeader(name: String): HttpMiddleware[Any, Nothing] =
+    patch((_, _) => Patch.removeHeaders(List(name)))
+
+  /**
+   * Runs the effect after the response is produced
+   */
+  def runAfter[R, E](effect: ZIO[R, E, Any]): HttpMiddleware[R, E] =
+    patchZIO((_, _) => effect.mapBoth(Option(_), _ => Patch.empty))
+
+  /**
+   * Runs the effect before the request is passed on to the HttpApp on which the middleware is applied.
+   */
+  def runBefore[R, E](effect: ZIO[R, E, Any]): HttpMiddleware[R, E] =
+    makeResponseZIO((_, _, _) => effect.mapError(Option(_)).unit)((_, _, _) => UIO(Patch.empty))
+
+  /**
+   * Creates a new middleware that always sets the response status to the provided value
+   */
+  def status(status: Status): HttpMiddleware[Any, Nothing] = patch((_, _) => Patch.setStatus(status))
+
+  /**
+   * Creates a new constants middleware that always executes the app provided, independent of where the middleware is
+   * applied
+   */
+  def fromApp[R, E](app: HttpApp[R, E]): HttpMiddleware[R, E] = Middleware.fromHttp(app)
+
+  /**
+   * Times out the application with a 408 status code.
+   */
+  def timeout(duration: Duration): HttpMiddleware[Clock, Nothing] =
+    Middleware.identity.race(Middleware.fromApp(Http.status(Status.REQUEST_TIMEOUT).delayAfter(duration)))
+
+  /**
+   * Creates a new middleware using a function from request parameters to a HttpMiddleware
+   */
+  def fromMiddlewareFunction[R, E](f: RequestP[HttpMiddleware[R, E]]): HttpMiddleware[R, E] =
+    Middleware.make(req => f(req.method, req.url, req.getHeaders))
+
+  /**
+   * Creates a new middleware using a function from request parameters to a ZIO of HttpMiddleware
+   */
+  def fromMiddlewareFunctionZIO[R, E](f: RequestP[ZIO[R, E, HttpMiddleware[R, E]]]): HttpMiddleware[R, E] =
+    Middleware.makeZIO[Request](req => f(req.method, req.url, req.getHeaders))
+}
+
+object HttpMiddlewares {
+
+  final case class PartialResponseMake[S](req: (Method, URL, Headers) => S) extends AnyVal {
+    def apply(res: (Status, Headers, S) => Patch): HttpMiddleware[Any, Nothing] = {
+      Middleware.intercept[Request, Response](
+        incoming = request => req(request.method, request.url, request.getHeaders),
+      )(
+        outgoing = (response, state) => res(response.status, response.getHeaders, state)(response),
+      )
+    }
+  }
+
+  final case class PartialResponseMakeZIO[R, E, S](req: (Method, URL, Headers) => ZIO[R, Option[E], S]) extends AnyVal {
+    def apply[R1 <: R, E1 >: E](res: (Status, Headers, S) => ZIO[R1, Option[E1], Patch]): HttpMiddleware[R1, E1] =
+      Middleware
+        .interceptZIO[Request, Response]
+        .apply[R1, E1, S, Response](
```

Comment:
Like we have done here we should do something similar with `interceptZIO` also.
Such that we don't need to specify the type params here.